content/Proposals/Done/202106-proposals-process.md

```diff
@@ -0,0 +1,112 @@
+* Existing Google doc-based documents are hard to consume. The formatting is widely different. Naming is inconsistent.
+* Document creation is rarely actionable. There is no review process, so the effort of creating a relevant document might be wasted, as the document is lost. This also leads to docs being in the half-completed state, demotivating readers to look at it.
+* It's hard to track previous discussions around proposals, who approved them (e.g. proposals).
+* It's not public, and it's hard to share good proposals with other external and internal teams.
+
+## Goals
+
+Goals and use cases for the solution as proposed in [How](#how):
+
+* Allow easy collaboration and decision making on design ideas.
+* Have a consistent design style that is readable and understandable.
+* Ensure design docs are discoverable for better awareness and knowledge sharing about past decisions.
+* Define a clear review and approval process.
+
+## Non-Goals
+
+* Define process for other documents (see [handbook proposal](../Accepted/202106-handbook.md#pitfalls-of-the-current-solution))
+
+## How
+
+We want to propose an improved, official proposal process for Monitoring Group that clearly states *when, where and how* to create proposal/enhancement/design documents.
+
+Everything starts with a problem statement. It might be a missing functionality, confusing existing functionality or broken one. It might be an annoying process, performance or security issue (or potential one).
+
+### What Process to Follow?
+
+As defined in [handbook proposal](../Accepted/202106-handbook.md#pitfalls-of-the-current-solution), our Handbook should tell you that Handbook is meant to be an index for our team resources and linking point to other distributed projects we maintain or contribute to.
+
+First, we need to identify if the idea we have is something we can contribute to the upstream project, or it does not fit anywhere else, so we can leverage the [Handbok Proposal directory](..) and the [process](#handbook-proposal-process). See the below algorithm to find it out:
+
+![where](../../assets/proposal-where.png)
+
+[Internal Team Drive for Public and Confidential Proposals](https://drive.google.com/drive/folders/1WGqC3gMCxIQlrnjDUYfNUTPYYRI5Cxto)
+
+### Handbook Proposal Process
+
+If there is no problem, there is no need for changing anything, no need for a proposal. This might feel trivial, but we should first ask ourselves this question before even thinking about writing a proposal.
+
+It takes time to propose an idea, find consensus and implement more significant concepts, so let's not waste time before it's worth it. But, unfortunately, even good ideas sometimes have to wait for a good moment to discuss them.
+
+Let's assume the idea sounds interesting to you; what to do next, where to propose it? How to review it? Follow the algorithm below:
+
+![where](../../assets/proposal-how.png)
```

Comment:
Is there an SLO on how long we want proposals to:
* Stay in Review state
* Stay in Approved state before starting implementation.

Reply:
I don't think so. It's similar to PR (code change) process now? Depends on sprint priority. Do you think we need SLO?

Reply:
I think there should be an SLO for how long things stay in review state but not necessarily how long before they are implemented. Our problem to-date has been that docs hang around in review-limbo for eternity, which is super demotivating for writers and means they don't get credit for their good ideas. Whether or not a design is implemented is a different problem: the design may be a great idea but the feature may not be prioritized by the PM so it's not on our roadmap for the coming X quarters. Architects and leads should be responsible for circulating designs to the right ppl, e.g. managers and PMs, and to advocate for prioritizing them as needed, but it's not only up to the team to decide what we work on. Otherwise, cynically, this SLO would be a backdoor to override the needs of Red Hat as a business and get engineer hours dedicated to pet projects that have architect approval.

Reply:
What specific SLO we should set?

Each iteration between change and review should not last longer than 3 days? 🤔

Reply:
I think it makes sense to set something that is simple to start with. We can always adjust it if we need to. E.g. we can say each proposal needs to be approved or rejected within 1 month. We can also easily measure this SLO to see whether we are breaching it. 

I guess the more important question is, what is the SLA for breaching the SLO :)

Reply:
@fpetkovski I guess if you breach the SLO then you lose the chance to voice your opinion and steer the proposal. It's a tricky one though because different proposals are going to have different scope and so might need lots/little time and or energy investment to review. Setting a low value for an SLO could be problematic in respect to timing for someone that might be interested in reviewing and have good opinions but be swamped etc?

I suppose at the very least if we do set an SLO, a comment from an interested party such as `I'm interested but don't have scope to review right now, I need more time!` would suffice, right?

Reply:
That's a good point indeed, makes total sense :+1:

Reply:
> I guess if you breach the SLO then you lose the chance to voice your opinion and steer the proposal

I'm thinking more about this from the other point of view, i.e. from that of the person proposing the document. If the SLO is breached, then it's them who suffer because their ideas won't be recognized. I think that in the last year this has been a bigger problem for the team than a reviewer not getting a say in the process.

Reply:
@squat that's a fair point. I was mostly thinking that in a team of ~20, the proposer should be able to get some reviews within that time frame and drive the idea forward without being blocked by someone in particular.

So then is the suggestion that whoever proposes the idea might be able to identify and cc one or more people and push the onus on them to review within a specific timeframe?

Reply:
I think that there needs to be some "official" reviewer who is accountable for approving or rejecting a proposal within x time. The problem IMO has not been that no one reads proposals but rather that they are read and they are never officially taken out of limbo. Even if you ping x person to review a doc, who is to say they are authoritative enough to give official sign-off on a proposal? For this reason it's tempting to default to saying architect/lead has to accept/reject a proposal within x weeks.

Reply:
@squat yeah I get your point, makes total sense and at the end of the day someone or some relativity small group need to take the lead and "steer the ship" so I am all for it in that respect.

Reply:
> The problem IMO has not been that no one reads proposals but rather that they are read and they are never officially taken out of limbo.

I agree that this is a problem, I am just not sure if approving a proposal and not acting on it makes things better. With the current team processes, sprint and roadmap planning is done by a handful of people. This makes it hard for proposal authors to get visibility into if and when their proposal would be planned for implementation. Approving a proposal and not implementing it for a year or two still feels like a limbo state.

It was mentioned above that approving and implementing a proposal should be treated as different problems. I think this could be a good way to look at things. However, not having a feedback loop for the implementation phase leaves a big gap in the process.

Reply:
I totally see this problem, but I think it's better to NOT couple the proposal process with the implementation process.

> I am just not sure if approving a proposal and not acting on it makes things better.

I have to disagree. (: Let's say we don't have time to implement a new idea which someone has in mind. Do we think it will help to NOT do the proposal at least? I feel it's extremely helpful to at least cover that idea while it's hot, so anyone can take it up anytime later. Especially when we are working open-source it helps the community to get it up for us too! Not mentioning we would have a ready answer if someone asks about new idea that was discussed before. We can point how we would discuss this and that person can either update proposal if needed with new info or see what was decided before.

Also we do the same for bugs no? We acknowledge them but sometimes we can't fix everything. Still it's good to report them instead of ignoring and move the fix process as much as further as possible.

> However, not having a feedback loop for the implementation phase leaves a big gap in the process.

In some way there is a process for planning things for our teams: This is part of backlog grooming, sprint planning, syncs with team leads. Those are not properly documented, and probably there should be another piece of documentation for that WDYT? (:

Reply:
Ok, this seems like a good direction. To summarize:
* Keep approval and implementation phases independent
* Potentially iterate on adding visibility into planning and grooming processes of the teams. I would not rush this work to much before the team stabilizes. Constantly updating documentation can be a bit wasteful :)

Reply:
Proposal - follow KEP process.
- Proposal itself should state who is the approver/reviewer (creator can assign on first draft and can be changed in further iterations)
- Proposal should state if it is implementable right now or if it isn't
- Proposals without any activity are labeled as `stale` after 60d of inactivity
- Stale proposals are closed after an additional 30d of inactivity

More in KEP-0000 - https://github.com/kubernetes/enhancements/blob/master/keps/sig-architecture/0000-kep-process/README.md